allennlp/data/dataset_readers/reading_comprehension/triviaqa.py

```diff
@@ -1,32 +1,245 @@
+    """
+    texts: List[str]
+    tokens: List[List[Token]]
+    token_spans: List[List[Tuple[int, int]]]
+    has_answers: List[int] = None
+
+    def to_json(self) -> JsonDict:
+        return {
+                "texts": self.texts,
+                "tokens": [[token_to_json(token) for token in tokens_i]
+                           for tokens_i in self.tokens],
+                "token_spans": self.token_spans,
+                "has_answers": self.has_answers
+        }
+
+    @staticmethod
+    def from_json(blob: JsonDict) -> 'MergedParagraphs':
+        return MergedParagraphs(texts=blob['texts'],
+                                tokens=[[json_to_token(b) for b in tokens_i]
+                                        for tokens_i in blob['tokens']],
+                                token_spans=blob['token_spans'],
+                                has_answers=blob.get('has_answers'))
+
+
+class Question(NamedTuple):
+    """
+    We process each TriviaQA question into tokenized question text
+    and merged paragraphs that may or may not contain the answer.
+    """
+    id: str  # pylint: disable=invalid-name
+    text: str
+    tokens: List[Token]
+    paragraphs: MergedParagraphs
+    answer_texts: List[str] = None
+
+    def to_json(self) -> JsonDict:
+        return {
+                "id": self.id,
+                "text": self.text,
+                "tokens": [token_to_json(token) for token in self.tokens],
+                "paragraphs": self.paragraphs.to_json(),
+                "answer_texts": self.answer_texts
+        }
+
+    @staticmethod
+    def from_json(blob: JsonDict) -> 'Question':
+        return Question(id=blob['id'],
+                        text=blob['text'],
+                        tokens=[json_to_token(token) for token in blob['tokens']],
+                        paragraphs=MergedParagraphs.from_json(blob['paragraphs']),
+                        answer_texts=blob.get('answer_texts'))
+
+
+def _merge(paragraphs: List[str],
+           question: str,
+           answer_texts: List[str],
+           tokenizer: Tokenizer,
+           max_paragraphs: int = 4,
+           tfidf_sort: bool = True,
+           require_answer: bool = False,
+           max_paragraph_length: int = 400) -> MergedParagraphs:
+    """
+    Given some (small?) paragraphs and a question, merge the paragraphs to make
+    synthetic paragraphs of size ``max_paragraph_length``, and return a ``MergedParagraphs``
+    instance containing the "best" ``topn`` of them. If ``topn`` is None, return all of them.
+
+    If ``tfidf_sort`` is True, "best" means closest (in terms of tfidf distance) to the question.
+    If ``require_answer`` is True, then the best "best" paragraph is also required to contain
+    one of the ``answer_texts``. Don't do this with a test or validation dataset, that's cheating.
+    If neither is True, then just the first ``topn`` paragraphs are returned.
+    """
+    # Collect all the text into one mega-paragraph.
```

Comment:
This is a problem.  You're losing all paragraph (and even sentence) boundaries, then splitting on number of tokens.  This means that you'll get paragraphs split in the middle of sentences.  This will certainly cause a drop in accuracy relative to Chris's implementation.  You can see Chris's original code here: https://github.com/allenai/document-qa/blob/35c9b862c5a743b11f9532b3e1ba11c348bb9621/docqa/data_processing/document_splitter.py#L302-L345.  Basically: he maintains paragraph boundaries, and just merges consecutive paragraphs until he reaches the max tokens; if adding the next paragraph would put you over the limit, you start a new paragraph.  If there's a single paragraph that's longer than the max tokens, he truncates it.